feat(web-analytics): Display values from previous period alongside current data

rafaeelaudibert opened the pull request:
## Problem

We want to be able to see at a glance if the current numbers are going up/down compared to the previous period. This came up in a user interview we did yesterday, and we thought it was an easy win.

## Changes

Queries are slightly heavier now, because we need to query for double the period we were doing previously and then split then using `*If` functions.

In the UI, we're simply representing this with a trend icon which displays more information on hover.\\
<img width="1358" alt="image" src="https://github.com/user-attachments/assets/74e0df21-0d07-4eb9-9963-b66133b88110">
<img width="1186" alt="image" src="https://github.com/user-attachments/assets/325fa24e-6513-4fb0-ad28-e3667d9817d4">


## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Unit tests for the backend, manual testing for the UI

rafaeelaudibert committed:
Start measuring previous values for our tables

All of our web analytics table now return the current/previous value to the UI, similar to the overview view. This will allow us to render a component clarifying if it's up or down from the past period

rafaeelaudibert committed:
Convert current/previous calculation to tuples

This is much easier to read and to test, it also greatly simplifies the UI

rafaeelaudibert committed:
Add new variation icons to the frontend

We can now easily see what the growth looks like period over period in the UI

rafaeelaudibert committed:
Fix double negative

We were saying "decreased by -10%" which doesnt make sense, let's just say "decreased by 10%" instead

rafaeelaudibert committed:
Use more generic types

rafaeelaudibert commented:
Don't you think the UI looks too noisy if this is turned on all the time?

robbie-c commented on posthog/hogql_queries/web_analytics/stats_table.py:
```diff
@@ -340,15 +397,17 @@ def to_path_bounce_query(self) -> ast.SelectQuery:
     SELECT
         breakdown_value,
-        avg(is_bounce) AS bounce_rate
+        coalesce(avgIf(is_bounce, current_period_segment), 0) AS bounce_rate,
```
Do we need the coalesce to zero? IMO returning NULL is more accurate

rafaeelaudibert replied:
I've changed this back to returning `NULL`

rafaeelaudibert committed:
Avoid displaying trend icon if previous value is non-existent

rafaeelaudibert committed:
Revert to computing `None` if bounce rate can't be computed (versus being 0)

The UI has also been updated to more correctly display icons, and avoid displaying trend icons if there's no old data

robbie-c commented:
Maybe icon size
Maybe icon padding
Padding when no arrow or always add arrow
Infinity percent
Tooltip triangle position

rafaeelaudibert committed:
Render infinity icon

rafaeelaudibert committed:
Fix non-matching width

rafaeelaudibert committed:
Fix tooltip triangle position

Wrap content in span to make sure the tooltip is over the middle of the text rather than the middle of the table cell

rafaeelaudibert commented:
@robbie-c Ready to rereview, worked on most of the changes you mentioned above

- Icon padding
<img width="395" alt="image" src="https://github.com/user-attachments/assets/4ed6cb23-1bfe-49ae-86fe-a7268ae5ebbd">


- Infinity percent + tooltip arrow in the right place
<img width="472" alt="image" src="https://github.com/user-attachments/assets/02ec0959-0ec5-4070-a3f0-9c960bb39120">

robbie-c commented on frontend/src/lib/utils.tsx:
```diff
@@ -124,6 +124,10 @@ export function percentage(
     maximumFractionDigits: number = 2,
     fixedPrecision: boolean = false
 ): string {
+    if (division === Infinity) {
```
```suggestion
    if (!isFinite(division)) {
```

robbie-c replied:
(js has positive and negative infinity)
